packages/loopring_v3/circuit/Gadgets/StorageGadgets.h

```diff
@@ -199,47 +213,54 @@ class StorageReaderGadget : public GadgetT {
   }
 };
 
-class NonceGadget : public GadgetT {
+// QUESTION(daniel): it seems this gadget is the same as ReadStorageGadget
+// except that it requires data to be 0.
+//
+// Maybe we should move this to `AccountGadgets.h' as this is on the account
+// level?
+class ReadNonceGadget : public GadgetT {
   const Constants &constants;
   const DualVariableGadget &storageID;
 
-  StorageReaderGadget storageReader;
+  ReadStorageGadget readStorage;
   IfThenRequireEqualGadget requireDataZero;
 
 public:
-  NonceGadget(ProtoboardT &pb, const Constants &_constants,
-              const StorageGadget &storage,
-              const DualVariableGadget &_storageID, const VariableT &verify,
-              const std::string &prefix)
-      : GadgetT(pb, prefix),
-
-        constants(_constants), storageID(_storageID),
-
-        storageReader(pb, constants, storage, storageID, verify,
-                      FMT(prefix, ".storageReader")),
-        requireDataZero(pb, verify, storageReader.getData(), constants._0,
-                        FMT(prefix, ".requireDataZero")) {}
+  ReadNonceGadget(ProtoboardT &pb, const Constants &_constants,
+              const StorageGadget &_currentStorage,
+              const DualVariableGadget &_storageID, const VariableT &_verify,
+              const std::string &_prefix)
+      : GadgetT(pb, _prefix),
+
+        constants(_constants), // QUESTION(daniel): can we not have the local
+                               // variables and
+        storageID(_storageID), // use _constant/_storageID directly?
+        readStorage(pb, constants, _currentStorage, storageID, _verify,
+                    FMT(_prefix, ".readStorage")),
+        requireDataZero(pb, _verify, readStorage.getData(), _constants._0,
+                        FMT(_prefix, ".requireDataZero")) {}
 
   void generate_r1cs_witness() {
-    storageReader.generate_r1cs_witness();
+    readStorage.generate_r1cs_witness();
     requireDataZero.generate_r1cs_witness();
   }
 
   void generate_r1cs_constraints() {
-    storageReader.generate_r1cs_constraints();
+    readStorage.generate_r1cs_constraints();
     requireDataZero.generate_r1cs_constraints();
   }
 
   const VariableT &getData() const { return constants._1; }
 
+  // QUESTION(daniel): need some hint to understand.
```

Comment:
With only the slot ID + if the data gets overwritten we can correctly reconstruct the Merkle tree with just this data in the DA. If it's overwritten than the data is reset to 0 and the `storageID` stored in the leaf is incremented by `numStorageSlots`. If it's not overwritten the `storageID` stored in the leaf remains the same and `data` increases either by 1 for transfers, or by the amount filled for trades.

So with just 2 bytes we have all we need. This helped more when `storageID` was much larger (8 bytes), but now the max `storageID` is only 4 bytes so only helps a little bit so save on DA.

Reply:
OK，I understand the nonce logics now. But still: what the return-value means for getShortStorageID()?

Reply:
It contains the bare minimum amount of data needed to reconstruct what happened in the circuits. So it's only used for DA, nowhere else.

The height of the storage tree is only 14, so with 14 bits we know exactly which leaf is being modified. But we don't know yet what the actual storageID is. But there can only be 2 cases: the storageID == leaf.storageID (in this case the overwrite bit is 0), or storageID == leaf.storageID + numSlots (in this case the overwrite bit is 1 because the data is reset before it is used).

But because the storageID is only 4 bytes now (back from 8 bytes like it was), maybe we could drop it altogether. Theoretically this adds quite a number of DA bytes to trades/transfers (+4 DA bytes for trades, +2 DA bytes for transfers), but in practice `storageID` bytes will be mostly small and will contain many zeros, so the cost reduction may be pretty small.